projects/igniteui-angular/src/lib/grids/toolbar/grid-toolbar-advanced-filtering.component.ts

```diff
@@ -38,5 +40,26 @@ export class IgxGridToolbarAdvancedFilteringComponent {
     @Input()
     public overlaySettings: OverlaySettings;
 
-    constructor( @Inject(IgxToolbarToken) private toolbar: IgxToolbarToken) { }
+    constructor( @Inject(IgxToolbarToken) private toolbar: IgxToolbarToken) {
+        this.grid?.advancedFilteringExpressionsTreeChange.subscribe(filteringTree => {
+            this.numberOfColumns = this.extractUniqueFieldNamesFromFilterTree(filteringTree).length;
+        });
+    }
+
+    public ngAfterViewInit(): void {
+        this.numberOfColumns = this.grid?.advancedFilteringExpressionsTree ? this.extractUniqueFieldNamesFromFilterTree(this.grid?.advancedFilteringExpressionsTree).length : 0;
+    }
+
+    protected extractUniqueFieldNamesFromFilterTree(filteringTree?: IFilteringExpressionsTree) : string[] {
+        const columnNames = [];
+        if (!filteringTree) return columnNames;
+        filteringTree.filteringOperands.forEach((expr) => {
+            if (expr instanceof FilteringExpressionsTree) {
+                columnNames.push(...this.extractUniqueFieldNamesFromFilterTree(expr));
+            } else {
+                columnNames.push((expr as IFilteringExpression).fieldName);
+            }
+        });
+        return columnNames.filter((value, index, self) => self.indexOf(value) === index);
```

Comment:
PS: Though I doubt those arrays will be big enough to make a difference but `[...new Set(columnNames)]` is shorter and should be faster :)